libs/cardano-data/src/Data/OMap/Strict.hs

```diff
@@ -0,0 +1,429 @@
+-- OMap {omSSeq = StrictSeq {fromStrict = fromList [2]}, omMap = fromList [(2,OMapTest {omFst = 2, omSnd = 1})]
+adjust :: HasOKey k v => (v -> v) -> k -> OMap k v -> OMap k v
+adjust f k omap@(OMap sseq kv) =
+  case Map.lookup k kv of
+    Nothing -> omap
+    Just v ->
+      let v' = f v
+          k' = v' ^. okeyL
+       in if k' == k
+            then OMap sseq (Map.insert k v' kv)
+            else
+              let kv' = Map.insert k' v' $ Map.delete k kv
+                  (lseq, rseq) = case SSeq.spanl (/= k) sseq of
+                    (l, _ SSeq.:<| r) -> (l, r)
+                    _ -> error "Impossible: supplied key expected to be in the sequence"
+               in case Map.lookup k' kv of
+                    Nothing -> OMap (lseq <> (k' SSeq.:<| rseq)) kv'
+                    Just _ -> OMap (lseq <> rseq) kv'
+
+-- | \(O(1)\)
+pattern Empty :: OMap k v
+pattern Empty <- (null -> True)
+  where
+    Empty = empty
+
+-- | \(O(\log n)\).
+pattern (:<|:) :: (HasOKey k v, Ord k) => v -> OMap k v -> OMap k v
+pattern x :<|: xs <- (uncons -> Just (x, xs))
+  where
+    x :<|: xs = x <| xs
+
+infixr 5 :<|:
+
+-- | \(O(\log n)\).
+pattern (:|>:) :: (HasOKey k v, Ord k) => OMap k v -> v -> OMap k v
+pattern xs :|>: x <- (unsnoc -> Just (xs, x))
+  where
+    xs :|>: x = xs |> x
+
+infixl 5 :|>:
+
+{-# COMPLETE Empty, (:|>:) #-}
+{-# COMPLETE Empty, (:<|:) #-}
+
+-- | \( O(n \log m) \). For every uncons-ed element from the sequence on the right,
+-- check its membership in the sequence on the left, before snoc'ing it.
+-- Preserve order. Remove duplicates from sequence on the right.
+(|><) :: HasOKey k v => OMap k v -> OMap k v -> OMap k v
+omapl |>< omapr = case omapr of
+  Empty -> omapl
+  r :<|: rs -> (omapl |> r) |>< rs
+
+infixl 5 |><
+
+-- | \( O(m \log n) \). For every unsnoc-ed element from the sequence on the left,
+-- check its membership in the sequence on the right, before cons'ing it.
+-- Preserve order. Remove duplicates from sequence on the left.
+(><|) :: HasOKey k v => OMap k v -> OMap k v -> OMap k v
+omapl ><| omapr = case omapl of
+  Empty -> omapr
+  ls :|>: l -> ls ><| (l <| omapr)
+
+infixr 5 ><|
+
+instance HasOKey k v => IsList (OMap k v) where
+  type Item (OMap k v) = v
+  fromList = fromFoldable
+  toList = F.toList
+
+instance (HasOKey k v, ToExpr v) => ToExpr (OMap k v) where
+  listToExpr = listToExpr . F.toList
+  toExpr = toExpr . F.toList
+
+instance (HasOKey k v, ToJSON v) => ToJSON (OMap k v) where
+  toJSON = toJSON . toStrictSeq
+  toEncoding = toEncoding . toStrictSeq
+
+instance HasOKey k v => Semigroup (OMap k v) where
+  (<>) = (|><)
+
+instance HasOKey k v => Monoid (OMap k v) where
+  mempty = empty
+
+instance Ord k => Foldable (OMap k) where
+  foldMap f = F.foldMap f . toStrictSeq
+  {-# INLINEABLE foldMap #-}
+  foldr f z = F.foldr f z . toStrictSeq
+  {-# INLINEABLE foldr #-}
+  foldl f z = F.foldl f z . toStrictSeq
+  {-# INLINEABLE foldl #-}
+  foldr' f z = F.foldr' f z . toStrictSeq
```

Comment:
Thanks for the awesome suggestions @lehins! I have tried the same for `foldr1` and `foldl1` without success however, since they both use a type signature that does not work for a similar implementation. `foldl1 :: (a -> a -> a) -> t a -> a`. It throws a `rigid type variable` error because in our implementation `k` and `v` are not expected to unify. Should we be skipping those two, so they get expressed automatically in terms of `foldr` and `foldl`?